# Introduction / Purpose

In [1]:
from csv import reader

# pull in google play dataset
openfile = open('data/googleplaystore.csv')
fileopen = reader(openfile)
#create dataset list object
android = list(fileopen)
#strip headers into own list and update android list object to have no headers
android_header = android[0]
android = android[1:]

# pull in apple store dataset
openfile = open('data/AppleStore.csv')
fileopen = reader(openfile)
# create dataset list object
ios = list(fileopen)
#strip headers into own list and update ios list object to have no headers
ios_header = ios[0]
ios = ios[1:]

# print samples to endure working
print('Android Sample: \n')
print(android_header)
print(android[:5])
print('\niOS Sample: \n')
print(ios_header)
print(ios[:5])

Android Sample: 

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']
[['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up'], ['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up'], ['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up'], ['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up'], ['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Fre

In order to visualise the data, explore_data will be created as a function. This will also contain a flag that can be used in order to count the rows and columns within a dataset

In [4]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]
    for row in dataset_slice:
        print(row)
        print('\n')
        
    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))
        
print(android_header)
print('\n')
explore_data(android, 0, 3, True)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


Number of rows: 10841
Number of columns: 13


The data shows that the google dataset contains 10841 rows and 13 columns. Glancing at the column headers, ones that may prove to be useful for analysis are 'App', 'Category', 'Reviews', 'Type', 'Price' and 'Genre'.

Next view same information but on the Apple dataset

In [5]:
print(ios_header)
print('\n')
explore_data(ios, 0, 3, True)

['', 'id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['1', '281656475', 'PAC-MAN Premium', '100788224', 'USD', '3.99', '21292', '26', '4', '4.5', '6.3.5', '4+', 'Games', '38', '5', '10', '1']


['2', '281796108', 'Evernote - stay organized', '158578688', 'USD', '0', '161065', '26', '4', '3.5', '8.2.2', '4+', 'Productivity', '37', '5', '23', '1']


['3', '281940292', 'WeatherBug - Local Weather, Radar, Maps, Alerts', '100524032', 'USD', '0', '188583', '2822', '3.5', '4.5', '5.0.0', '4+', 'Weather', '37', '5', '3', '1']


Number of rows: 7197
Number of columns: 17


The Apple dataset contains 7197 rows and 17 columns. Headers of interest for the analysis are 'track_name', 'currency', 'price', 'rating_count_tot', 'rating_count_ver' and 'prime_genre'

# Data Processing

Reviewing the discussion forums on Kaggle for where the google dataset was accertained, one of the discussions mention mising data on row 10472. The discussion does not stipulate if row 10472 is when the dataset contains headers or not so the first port of call is to verify if we having missing data in that row or surrounds rows.

In [8]:
explore_data(android, 10471, 10474)
print(android_header)

['Xposed Wi-Fi-Pwd', 'PERSONALIZATION', '3.5', '1042', '404k', '100,000+', 'Free', '0', 'Everyone', 'Personalization', 'August 5, 2014', '3.0.0', '4.0.3 and up']


['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


['osmino Wi-Fi: free WiFi', 'TOOLS', '4.2', '134203', '4.1M', '10,000,000+', 'Free', '0', 'Everyone', 'Tools', 'August 7, 2018', '6.06.14', '4.4 and up']


['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


The above shows that missing data is present on row 10472 and that the missing data refers to the genre of the app titled Life Made WI-Fi Touchscreen Photo Frame.
The missing data from this data could skrew the analysis of the data, it will therefore be removed from the dataset. To ensure that i. the row is deleted and ii. that only a single row is removed, the dataset row count will be printed either side of the deletion.

In [9]:
print('No. of rows before deletion:', len(android))
del android[10472]
print('No. of rows after deletion:', len(android))

No. of rows before deletion:  10841
No. of rows after deletion 10840


# Duplicated Entries

The forums on the Google dataset claim that there are duplicated app entries within the dataset, where as in the Apple dataset there is none. We will verify the duplications in the Google dataset first before formulating a solution for there removal.

In [12]:
duplicated_apps = []
unique_apps = []

for app in android:
    name = app[0]
    if name in unique_apps:
        duplicated_apps.append(name)
    else:
        unique_apps.append(name)
        
print('Number of duplicated apps:', len(duplicated_apps))
print('\n')
print('Sample of duplicated apps:', duplicated_apps[:5])

Number of duplicated apps: 1181


Sample of duplicated apps: ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings']


The data shows that we have 1181 duplicated apps in the dataset that can skew the analysis, these will need to be removed keeping a single instance of the app. Before this can be done, the duplicates of a single app need to be inspected to see if the data is identical or if different.

In [14]:
for app in android:
    name = app[0]
    if name == 'Instagram':
        print(app)

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


The data shows a variation in the number of reviews for Instagram. We will therefore assume the same for other duplicates. 

The next steps to be taken are:

Create a dictionary of unique app names and the highest amount of reviews as a key value pair.
use the dictionary to build a new dataset where the app names are all unique.

In [15]:
max_reviews = {}

for app in android:
    name = app[0]
    n_reviews = float(app[3])
    if name in max_reviews and max_reviews[name] < n_reviews:
        max_reviews[name] = n_reviews
    elif name not in max_reviews:
        max_reviews[name] = n_reviews

Previously we noted 1181 duplicated apps, we will therefor confirm the functionality above by checking the android dataset length less the duplicated number is equal to the newly formed dictionary max_reviews

In [17]:
print('Expected length:', len(android) -1181)
print('Actual length:', len(max_reviews))

Expected length: 9659
Actual length: 9659


As the values are matching we can use max_reviews to create a new dataset ensuring no duplicate values are added and that out of the apps that are currently duplicated, only the one with the highest reviews are inserted into the new dataset

In [18]:
android_clean = []
already_added = []

for app in android:
    name = app[0]
    n_reviews = float(app[3])
    if (max_reviews[name] == n_reviews) and (name not in already_added):
        android_clean.append(app)
        already_added.append(name)

Quick check to ensure that the new dataset contains 9659 rows as expected

In [19]:
explore_data(android_clean, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9659
Number of columns: 13


# Clean Data

The datasets contain apps in different languages. In this instace we are only interested in English apps. The next step is to create a function that filters a string for ascii values, if we identify 3 chars over 127 then we can assume that the app in not english.

In [22]:
def is_english(string):
    non_ascii = 0
    
    for character in string:
        if ord(character) > 127:
            non_ascii += 1
        
    if non_ascii > 3:
        return False
    else:
        return True

I have chosen to look for 3 non ascii characters before identifing it as non english to prevent data loss. Some apps may contain emoji's or other symbols such as TM subscript that falls outside of the ascii 127. Next to test the new function

In [23]:
print('Expected results: True, False, True, True')
print(is_english('Instagram'))
print(is_english('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(is_english('Docs To Go™ Free Office Suite'))
print(is_english('Instachat 😜'))

Expected results: True, False, True, True
True
False
True
True


The results from the function are as expected. It is not a perfect solution but will limit the number of none English apps that make it though if any.

Now to use the function to clean the datasets further

In [26]:
android_english = []
ios_english = []

for app in android_clean:
    name = app[0]
    if is_english(name):
        android_english.append(app)

for app in ios:
    name = app[2]
    if is_english(name):
        ios_english.append(app)
        
#explore the results
explore_data(android_english, 0, 3, True)
print('\n')
explore_data(ios_english, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9614
Number of columns: 13


['1', '281656475', 'PAC-MAN Premium', '100788224', 'USD', '3.99', '21292', '26', '4', '4.5', '6.3.5', '4+', 'Games', '38', '5', '10', '1']


['2', '281796108', 'Evernote - stay organized', '158578688', 'USD', '0', '161065', '26', '4', '3.5', '8.2.2', '4+', 'Productivity', '37', '5', '23', '1']


['3', '281940292', 'WeatherBug - Local Weather, Radar, Maps, Alerts', '100524032', 'USD', '0', '188

We can now see a reduction in both datasets

# Filter out paid apps

As the organisation only builds free apps and makes revenue via ads, we are only interested in analysing the apps in the datasets that are free. The datasets contain both paid and free apps so need to be filtered down.

In [33]:
android_final = []
ios_final = []

for app in android_english:
    price = app[7]
    if price == '0':
        android_final.append(app)
        
for app in ios_english:
    price = app[5]
    if price == '0':
        ios_final.append(app)
        
print('Android Rows:', len(android_final))
print('Apple Rows:', len(ios_final))

Android Rows: 8864
Apple Rows: 3222


We now have two datasets that have been cleaned and are ready to analyse

# Most common apps by Genre

In order to view that data of most common apps by genre, two functions are required to show that data in a structured way (Frequency Table) and to also order the data correctly.

In [36]:
def freq_table(dataset, index):
    table = {}
    total = 0
    
    for row in dataset:
        total +=1
        value = row[index]
        if value in table:
            table[value] += 1
        else:
            table[value] = 1
            
    table_percentages = {}
    for key in table:
        percentage = (table[key] / total) * 100
        table_percentages[key] = percentage
        
    return table_percentages

def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_value_tuple = (table[key], key)
        table_display.append(key_value_tuple)
        
    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])
        

Examine the ios dataset based on the prime_genre column

In [38]:
display_table(ios_final, -5)

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


The data shows that games dominate the free app market on ios followed by entertainment, photo and video but the margin is hugely different. Intrestingly education apps only make up 3.6% of the free app space. Although these figures do show the frequency by genre, it is not taking into account the number of users which is a key success matrix for the organisation.

With that being said, lets inspect the android dataset

In [39]:
display_table(android_final, 1)

FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072
BUSINESS : 4.591606498194946
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2378158844765346
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.7937725631768955
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.1620036101083033
ENTERTAINMENT : 0.9589350180505415
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.6543321299638989
ART_AND_DESIGN : 